core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -306,6 +306,14 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       .index("endTime").reverse())(_.toApplicationInfo()).iterator
   }
 
+  override def getListing(max: Int)(predicate: ApplicationInfo => Boolean)
+    : Iterator[ApplicationInfo] = {
+    // Return the filtered listing in end time descending order.
+    KVUtils.mapToSeqWithFilter(
+      listing.view(classOf[ApplicationInfoWrapper]).index("endTime").reverse(),
+      max)(_.toApplicationInfo())(predicate).iterator
```

Comment:
Would using the semantics of `nonEmpty` make it clearer?

Reply:
> I am not sure what you mean by `nonEmpty` semantics @LuciferYang. Did you mean this for `shouldDisplayApplications` below ?

It's my fault. I wrote the comments in the wrong place. Your understanding is correct.